### DISCUSSION TITLE
How to contribute to pybricks project?

### DISCUSSION THREAD

**Discussion by nothingtodowu** [pybricks/support#1643] [General]
I just use pybricks for Lego Spike Prime and I think this is a great project for kids education and competitions. I want do some contribution for this project. My background is a WRO/FLL coach at Singapore and I have 25+ years firmware/linux driver development. I have EV3 and Spike Prime. For pybricks project I can do some testing task or some simple develop task. Can I contribute to pybricks project? Thanks a lot.

**Comment by laurensvalk** [comment]
Thank you for offering to help! That is much appreciated. We've been looking at different ways to continue support for EV3.  It would be awesome if you have some ideas or want to look at some of the existing ideas.

Perhaps you have seen that an older version of Pybricks exists for EV3 on top of ev3dev (debian) Linux. It only works with VS Code, not with the online editor. It also has a few other important issues, like loop time inconsistency. Perhaps most of all, it is quite different from the other hubs, which makes it harder to maintain going forward. This is why it is still at version 2.0.

One of the options we're considering is a version of Pybricks for EV3 _without_ Linux. By making it more like the other "bare metal" hubs, this may make it easier to maintain in the long run, and avoid some of the timing issues we had on the Linux version.

Then it can use the same code as on the other hubs, but with a different hardware abstraction layer for the AM1808 and other on board peripherals.

In a previous approach, we tried to achieve it with `ev3rt`, which is also included in `pybricks-micropython` repository. While this shows that everything can work (and very fast) it is a very elaborate code base (and uses TOPPERS RTOS), most of which we don't need. It seems quite hard to untangle.

So more recently, we've looked at a more minimal implementation in https://github.com/pybricks/pybricks-micropython/pull/241. This also includes references to various other projects. Since you have both the EV3 and extensive firmware/driver development experience, this could offer some interesting things to look at. What do you think?

**Reply by nothingtodowu** [reply]
Sound it is an exciting project. Since I am new to pybricks, maybe I can start from basic task. I have experience at Linux I2C/UART/QSPI FLASH driver. if any develop task or test task I can try. Thanks a lot.
